### DISCUSSION TITLE
How can I get good performance when drawing thousands of objects

### DISCUSSION THREAD

**Discussion by DrDryg** [AvaloniaUI/Avalonia#13042] [General]
I would like to make a control that draws a graph with up to many thousand nodes and a few times more connections than nodes. My target is that it should work smoothly with ideally 10000 nodes. (It is a bit of a overkill, but would guarantee that it works fine on less powerful hardware too.)

Each node would contain a few text strings, and a few primitives like squares and circles as well as same small icons. Parts of the node should be responsible to clicks and mouse-over. The connections will be lines or bezier curves. 

I've tested to make the nodes as user controls with labels, a grid, a stackpanel a few primitive objects like circles. However this becomes slow with a few hundred nodes. So my questions here are: 
- What would be a good approach to speed this up? 
- What performance can I expect, is 10000 objects unrealistic or achievable?   
- Should I draw the entire graph on a SkiaCanvas or can I use thousands of Avalonia controls?

In most cases the graph will be much smaller than 10000 objects and often if it is big one will not view it all at once. But a target of 10000 would guarantee that the UI is performant enough for all my use-cases with some margin.

**Comment by wieslawsoltes** [comment]
You can try approach taken in https://github.com/wieslawsoltes/Draw2D

Some example: https://github.com/wieslawsoltes/Draw2D/blob/master/src/Draw2D/Skia/Renderers/SkiaShapeRenderer.cs

**Reply by DrDryg** [reply]
Can you say a few words about the strategy used by Draw2D? It is quite a large project so would help me understand the codebase faster.

**Reply by timunie** [reply]
the idea should be to do: 
- some caching if possible
- Render on your own instead of using 10000 controls
- probably use a RenderTargetBitmap
- probably only draw points currently visible
